test/functional/runner/index.js

```diff
@@ -0,0 +1,591 @@
+        name: path.basename(x, '.json')
+      })
+    );
+}
+
+function parseTopologies(topologies) {
+  if (topologies == null) {
+    return ['replicaset', 'mongos', 'single'];
+  }
+
+  const idx = topologies.indexOf('single');
+  if (idx !== -1) {
+    topologies.splice(idx, 1);
+  }
+
+  return topologies;
+}
+
+function parseRunOn(runOn) {
+  return runOn.map(config => {
+    const topology = parseTopologies(config.topology);
+    const version = [];
+    if (config.minServerVersion) {
+      version.push(`>= ${config.minServerVersion}`);
+    }
+
+    if (config.maxServerVersion) {
+      version.push(`<= ${config.maxServerVersion}`);
+    }
+
+    const mongodb = version.join(' ');
+
+    return { topology, mongodb };
+  });
+}
+
+function shouldSkipTest(spec) {
+  const SKIP_TESTS = [
+    // commitTransaction retry seems to be swallowed by mongos in these three cases
+    'commitTransaction retry succeeds on new mongos',
+    'commitTransaction retry fails on new mongos',
+    'unpin after transient error within a transaction and commit',
+    'count'
+  ];
+
+  if (spec.skipReason || SKIP_TESTS.indexOf(spec.description) !== -1) {
+    return it.skip;
+  }
+
+  return it;
+}
+
+function generateTopologyTests(testSuites, testContext) {
+  testSuites.forEach(testSuite => {
+    const environmentRequirementList = parseRunOn(testSuite.runOn);
+
+    environmentRequirementList.forEach(requires => {
+      const suiteName = `${testSuite.name} - ${requires.topology.join()}`;
+      describe(suiteName, {
+        metadata: { requires },
+        test: function() {
+          beforeEach(() => prepareDatabaseForSuite(testSuite, testContext));
+          afterEach(() => testContext.cleanupAfterSuite());
+
+          testSuite.tests.forEach(spec => {
+            const maybeSkipIt = shouldSkipTest(spec);
+            maybeSkipIt(spec.description, function() {
+              let testPromise = Promise.resolve();
+
+              if (spec.failPoint) {
+                testPromise = testPromise.then(() => testContext.enableFailPoint(spec.failPoint));
+              }
+
+              // run the actual test
+              testPromise = testPromise.then(() =>
+                runTestSuiteTest(this.configuration, spec, testContext)
+              );
+
+              if (spec.failPoint) {
+                testPromise = testPromise.then(() => testContext.disableFailPoint(spec.failPoint));
+              }
+
+              return testPromise.then(() => validateOutcome(spec, testContext));
+            });
+          });
+        }
+      });
+    });
+  });
+}
+
+// Test runner helpers
+function prepareDatabaseForSuite(suite, context) {
+  context.dbName = suite.database_name;
+  context.collectionName = suite.collection_name;
+
+  const db = context.sharedClient.db(context.dbName);
+  const coll = db.collection(context.collectionName);
+
+  return db
+    .admin()
+    .command({ killAllSessions: [] })
+    .catch(() => {}) // ignore any error from this
```

Comment:
Why do we want to ignore errors here?

Reply:
its part of the test runner [specification](https://github.com/mongodb/specifications/blob/master/source/transactions/tests/README.rst)

Reply:
I think the spec says to only ignore error code 11601

Reply:
changed!